### PR TITLE
[C-runtime] Adding CVODE

### PR DESCRIPTION
**Pull request by AnHeuermann**
#  Ticket5830 CVODE in C runtime

Adding Sundials CVODE to the C runtime as requested in [ticket 5830](https://trac.openmodelica.org/OpenModelica/ticket/5830).

Similar to the way we use IDA solver.
More options and settings possible than in C++ runtime.

## New files
CVODE implementation:
  - cvode_solver.c
  - cvode_solver.h

Test files:
  - problem1-cvode.mos 
  - problem6-cvode.mos 

## Changes
- Added check in `printSparseStructure` to check for `NULL`
- Added CVODE to solver_main.c
- Fixed warnings during compilation of some `infoStreamPrint` calls
- Added new flags to choose CVODE solver and set some settings

## How to use
Use method="cvode" to integrate with CVODE with C runtime.
Use `-cvodeLinearMultistepMethod`  and `-cvodeNonlinearSolverIteration` to change integrator method.
Usual combinations:
  - `-cvodeLinearMultistepMethod=CV_BDF -cvodeNonlinearSolverIteration=CV_NEWTON` for stiff problems
  - `-cvodeLinearMultistepMethod=CV_ADAMS -cvodeNonlinearSolverIteration=CV_FUNCTIONAL` for non-stiff problems.

### CODE REVIEW

**Comment by AnHeuermann** [comment]
@kabdelhak Please note that all flag settings are in it's own function `void cvodeGetConfig(CVODE_CONFIG* config, threadData_t* threadData) ` which gets called during initialization of CVODE.

For FMU flags we can create an additional function next to it which will read settings from some file or could get them from the main initialization routine.
Or `cvode_solver_initial` will get some pointer to `cvodeGetConfig` or `cvodeGetConfigFMU` function and will use this function for optional settings.

**Comment by AnHeuermann** [comment]
One remaining problem is that for the bouncing ball example
```modelica
model problem6
  parameter Real e=0.7 "coefficient of restitution";
  parameter Real g=9.81 "gravity acceleration";
  Real h(start=1) "height of ball";
  Real v "velocity of ball";
  Boolean flying(start=true) "true, if ball is flying";
  Boolean impact;
  Real v_new;
  discrete Integer n_bounce(start=0);
equation
  impact = h <= 0.0;
  der(v) = if flying then -g else 0;
  der(h) = v;

  when {h <= 0.0 and v <= 0.0,impact} then
    v_new = if edge(impact) then -e*pre(v) else 0;
    flying = v_new > 0;
    reinit(v, v_new);
    n_bounce=pre(n_bounce)+1;
  end when;
end problem6;
```
I don't get the correct result for `v` respectively `der(h)`.

In Red: CVODE solution by C++ runtime
In Blue: CVODE solution by my new implementation
image


It's not a problem with the tolerance(making it smaller doesn't solve it) and not the scaling (this model is not scaled). Also the C++ version only uses the numerical approximation for the Jacobian provided by CVODE, so that should be fine as well.

Maybe I need to check the root finding function but I can't see the problem.
@lochel @kabdelhak  Any idea what could be the problem? Give me a lucky guess ;-)

**Comment by kabdelhak** [comment]
> One remaining problem is that for the bouncing ball example
> 
> ```modelica
> model problem6
>   parameter Real e=0.7 "coefficient of restitution";
>   parameter Real g=9.81 "gravity acceleration";
>   Real h(start=1) "height of ball";
>   Real v "velocity of ball";
>   Boolean flying(start=true) "true, if ball is flying";
>   Boolean impact;
>   Real v_new;
>   discrete Integer n_bounce(start=0);
> equation
>   impact = h <= 0.0;
>   der(v) = if flying then -g else 0;
>   der(h) = v;
> 
>   when {h <= 0.0 and v <= 0.0,impact} then
>     v_new = if edge(impact) then -e*pre(v) else 0;
>     flying = v_new > 0;
>     reinit(v, v_new);
>     n_bounce=pre(n_bounce)+1;
>   end when;
> end problem6;
> ```
> 
> I don't get the correct result for `v` respectively `der(h)`.
> 
> In Red: CVODE solution by C++ runtime
> In Blue: CVODE solution by my new implementation
> image
> 
> It's not a problem with the tolerance(making it smaller doesn't solve it) and not the scaling (this model is not scaled). Also the C++ version only uses the numerical approximation for the Jacobian provided by CVODE, so that should be fine as well.
> 
> Maybe I need to check the root finding function but I can't see the problem.
> @lochel @kabdelhak Any idea what could be the problem? Give me a lucky guess ;-)

Do you use regular event handling by bisection or do you use the CVODE event handling? Looks like it has something on its own?

File: testsuite/simulation/modelica/solver/problem6-cvode.mos [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // name: problem6-cvode
 // status: correct
-// teardown_command: rm -f testSolver.problem2* output.log
+// teardown_command: rm -f testSolver.problem2* output.log testSolver_problem6_*
```
Comment (lochel) [inline testsuite/simulation/modelica/solver/problem6-cvode.mos:3]: problem6?

File: testsuite/simulation/modelica/solver/Makefile [hunk 2]
```diff
@@ -19,6 +20,7 @@ problem2-trapezoid.mos \
 problem2-imprk.mos \
 problem2-irksco.mos \
 problem2-ida.mos \
+problem2-cvode.mos \
```
Comment (lochel) [inline testsuite/simulation/modelica/solver/Makefile:23]: problem6-cvode.mos?

**Comment by lochel** [comment]
@AnHeuermann It looks like a problem with the initialization.

**Comment by lochel** [comment]
I cannot reproduce your results. Here are my results for cvode-experimental:
image

File: OMCompiler/SimulationRuntime/c/util/simulation_options.c [hunk 3]
```diff
@@ -669,6 +687,7 @@ const char *SOLVER_METHOD_NAME[S_MAX] = {
   /* S_IRKSCO */        "irksco",
   /* S_DASSL */         "dassl",
   /* S_IDA */           "ida",
+  /* S_CVODE */         "cvode-experimental",
```
Comment (lochel) [inline OMCompiler/SimulationRuntime/c/util/simulation_options.c:690]: Please rename it to `cvode`. You can display a warning or notification to make people aware that this is a new feature. However, changing the name later seems a rather bad idea.

**Comment by kabdelhak** [comment]
> I cannot reproduce your results. Here are my results for cvode-experimental:
> image

I think it is the same, it is zoomed in. Look at the number of bounces... the normal runtime has it at about 33 for 3 minutes and with CVODE it is at over 2k

**Comment by lochel** [comment]
But that is fine. The model is bad and that is about numerics.

**Comment by kabdelhak** [comment]
> But that is fine. The model is bad and that is about numerics.


> But that is fine. The model is bad and the differences are caused by the numerics. Go ahead and merge this PR as soon as possible grinning


Yea sure, but it is weird that for all other solvers it converges at some point, only for CVODE it seems to stay bouncing forever. 

but i am all for merging :D

File: testsuite/simulation/modelica/solver/problem6-cvode.mos (hunk 1 above)
Comment (AnHeuermann) [reply testsuite/simulation/modelica/solver/problem6-cvode.mos:3]: Updated

File: testsuite/simulation/modelica/solver/Makefile (hunk 2 above)
Comment (AnHeuermann) [reply testsuite/simulation/modelica/solver/Makefile:23]: Added

File: OMCompiler/SimulationRuntime/c/util/simulation_options.c (hunk 3 above)
Comment (AnHeuermann) [reply OMCompiler/SimulationRuntime/c/util/simulation_options.c:690]: Users can't see a warning in OMEdit when choosing the integrator method, only after simulation, and I think changing the solver name is a minor thing.

Comment (lochel) [reply OMCompiler/SimulationRuntime/c/util/simulation_options.c:690]: OMEdit shows already a tooltip for the option which labels it as experimental and we can have a notification/warning when simulating. Renaming flags is always a very bad thing.

Comment (AnHeuermann) [reply OMCompiler/SimulationRuntime/c/util/simulation_options.c]: I renamed it

**Comment by casella** [comment]
@AnHeuermann [this report](https://libraries.openmodelica.org/branches/history/newInst/2020-05-29%2011:31:44..2020-05-29%2020:31:47.html) shows lots of regressions. I guess they are due to @perost's commit, but I'm not sure. Can you please check?

**Comment by AnHeuermann** [comment]
I would agree that it's cb30381695c9f4c10576388428eec82d386ce1cf.
Most of the failing models fail during compilation / template phase. My changes are at runtime level ,way after, and not active per default.

But I did manage to break OS-X (again....). See [ticket 5830](https://trac.openmodelica.org/OpenModelica/ticket/5830#comment:6). Don't know why though. It's happily building on Jenkins and on omodmac.ida.liu.se.

**Comment by casella** [comment]
We now have the [test results with -s=cvode and -d=newInst](https://libraries.openmodelica.org/branches/overview-cvode.html). Compare with the [standard results](https://libraries.openmodelica.org/branches/overview-newinst.html)

**Comment by casella** [comment]
Some cases fail because of duplicate -s settings, see e.g. [ScalableTestSuite](https://libraries.openmodelica.org/branches/cvode/ScalableTestSuite_noopt/ScalableTestSuite_noopt.html) models such as [this one](https://libraries.openmodelica.org/branches/cvode/ScalableTestSuite_noopt/files/ScalableTestSuite_noopt_ScalableTestSuite.Electrical.TransmissionLine.ScaledExperiments.TransmissionLineEquations_N_10.sim)

**Comment by casella** [comment]
In general I see many more issues with the [MSL 3.2.3](https://libraries.openmodelica.org/branches/cvode/Modelica_3.2.3/Modelica_3.2.3.html) compared to the [regular MSL 3.2.3 tests](https://libraries.openmodelica.org/branches/newInst/Modelica_3.2.3/Modelica_3.2.3.html), @AnHeuermann please investigate!

**Comment by casella** [comment]
Seems to me that the performance is similar in those cases that work, but I haven't carried out any systematic comparison.

**Pull request merged**